compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1166,26 +1168,38 @@ private BAttachedFunction getNextFunc(BObjectType iteratorType) {
 
     public BType inferRecordFieldType(BRecordType recordType) {
         List<BField> fields = recordType.fields;
+        BUnionType unionType = BUnionType.create(null);
 
-        // If there are no fields in the record, return the rest field type as the inferred type.
-        if (fields.isEmpty()) {
-            return recordType.restFieldType;
+        if (recordType.restFieldType.tag != TypeTags.NONE) {
+            unionType.add(recordType.restFieldType);
         }
 
-        BType inferredType = fields.get(0).type; // If all the fields are the same, doesn't matter which one we pick
+        for (BField field : fields) {
+            boolean fieldAssignableToUnionType = isAssignable(field.type, unionType);
+            boolean unionTypeAssignableField = isAssignable(unionType, field.type);
 
-        // If it's an open record, the rest field type should also be of the same type as the mandatory fields.
-        if (!recordType.sealed && recordType.restFieldType.tag != inferredType.tag) {
-            return symTable.anyType;
-        }
+            if (fieldAssignableToUnionType) {
```

Comment:
I guess we don't need a variable for `fieldAssignableToUnionType` since this is the only usage?